tools/capture-test.py

```diff
@@ -0,0 +1,373 @@
+# occasional constructed integer or byref() pointer.
+class ALSA:
+    PCM_STREAM_PLAYBACK = 0
+    PCM_STREAM_CAPTURE = 1
+    PCM_FORMAT_S16_LE = 2
+    PCM_FORMAT_S32_LE = 10
+    PCM_ACCESS_MMAP_INTERLEAVED = 0
+    PCM_ACCESS_RW_INTERLEAVED = 3
+    def __init__(self):
+        self.lib = c.cdll.LoadLibrary("libasound.so.2")
+    def __getattr__(self, name):
+        if name.startswith("snd_"):
+            fn = getattr(self.lib, name)
+            if name.endswith("_name"): # These return strings!
+                fn.restype = c.c_char_p
+                return lambda *args: fn(*args).decode("utf-8")
+            else:
+                return lambda *args: self.err_wrap(fn(*args))
+    def err_wrap(self, ret):
+        if ret < 0 and ret > -200:
+            raise OSError(os.strerror(-ret))
+        return ret
+    def alloc(self, typ):
+        return (c.c_byte * getattr(self.lib, f"snd_{typ}_sizeof")())()
+    class pcm_channel_area_t(c.Structure):
+        _fields_ = [("addr", c.c_ulong), ("first", c.c_int), ("step", c.c_int)]
+
+def pcm_init_stream(pcm, rate, chans, fmt, access):
+    hwp = alsa.alloc("pcm_hw_params")
+    alsa.snd_pcm_hw_params_any(pcm, hwp)
+    alsa.snd_pcm_hw_params_set_format(pcm, hwp, fmt)
+    alsa.snd_pcm_hw_params_set_channels(pcm, hwp, chans)
+    alsa.snd_pcm_hw_params_set_rate(pcm, hwp, rate, alsa.PCM_STREAM_PLAYBACK)
+    alsa.snd_pcm_hw_params_set_access(pcm, hwp, access)
+    alsa.snd_pcm_hw_params(pcm, hwp)
+
+# Noise reduction likes to squash our chirp on capture.  Walk the list
+# of controls, looking for an RTNR enable control, if one exists, and
+# set it to false.  Unbelievably cumbersome API to do this: call
+# elem_list once on an empty struct to get the element count, then
+# allocate, then call it again.  Then for each element we can check
+# the name directly, but need to allocate an "id" struct to query an
+# abstract identifier, that we use with a separately-allocated "value"
+# (on which we set the dyncmically typed data) to send the command to
+# the kernel.
+def ctl_disable_rtnr():
+    dev = f"hw:{opts.card}".encode("ascii")
+    ctl = c.c_ulong()
+    alsa.snd_ctl_open(c.byref(ctl), dev, 0)
+    elist = alsa.alloc("ctl_elem_list")
+    alsa.snd_ctl_elem_list(ctl, elist)
+    nelem = alsa.snd_ctl_elem_list_get_count(elist)
+    alsa.snd_ctl_elem_list_alloc_space(elist, nelem)
+    alsa.snd_ctl_elem_list(ctl, elist)
+    for i in range(nelem):
+        name = alsa.snd_ctl_elem_list_get_name(elist, i)
+        if re.match(r'RTNR.*\s+rtnr_enable.*', name):
+            print(f"Disabling control: {name}")
+            eid = alsa.alloc("ctl_elem_id")
+            val = alsa.alloc("ctl_elem_value")
+            alsa.snd_ctl_elem_list_get_id(elist, i, c.byref(eid))
+            alsa.snd_ctl_elem_value_set_id(val, eid)
+            alsa.snd_ctl_elem_value_set_boolean(val, 0, False)
+            alsa.snd_ctl_elem_write(ctl, val)
+    alsa.snd_ctl_close(ctl)
```

Comment:
not following how all other controls are initialized? The commit message states 'dependency-free ALSA' but the control parts will use whatever UCM or alsactl restored.

Reply:
That's exactly right.  You can set it up offline or reload the driver.  Honestly that's a feature, not a bug.  I want to exercise this code in different circumstances.  The RTNR stuff is just a workaround that was important enough to include directly.  Longer term, sure, it would be nice to have it grow more capability.

Reply:
> Honestly that's a feature, not a bug. I want to exercise this code in different circumstances.

Then this is an IMHO important part of the user interface. Please add one sentence about that in the HELP_TEXT.